### ISSUE TITLE
node get DiskPressure when disk space is still plenty

### ISSUE THREAD

**Issue by tingweiwu** [kubernetes/kubernetes#66961]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug

**What happened**:
the status of kubelet show DiskPressure 
```shell
[root@gpu16c-0001 overlay2]# systemctl status kubelet -l
● kubelet.service - kubelet: The Kubernetes Node Agent
   Loaded: loaded (/etc/systemd/system/kubelet.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/kubelet.service.d
           └─10-kubeadm.conf
   Active: active (running) since Fri 2018-08-03 16:12:26 CST; 50min ago
     Docs: http://kubernetes.io/docs/
 Main PID: 18808 (kubelet)
   Memory: 50.8M
   CGroup: /system.slice/kubelet.service
           └─18808 /usr/bin/kubelet --bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf --pod-manifest-path=/etc/kubernetes/manifests --allow-privileged=true --network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin --cluster-dns=10.96.0.10 --cluster-domain=cluster.local --authorization-mode=Webhook --client-ca-file=/etc/kubernetes/pki/ca.crt --cadvisor-port=0 --cgroup-driver=cgroupfs --rotate-certificates=true --cert-dir=/var/lib/kubelet/pki --feature-gates=DevicePlugins=true

Aug 03 17:02:43 gpu16c-0001.novalocal kubelet[18808]: W0803 17:02:43.866083   18808 eviction_manager.go:142] Failed to admit pod kube-proxy-z46zm_kube-system(ea8815fc-96fb-11e8-80c4-fa163eb0739b) - node has conditions: [DiskPressure]
Aug 03 17:02:44 gpu16c-0001.novalocal kubelet[18808]: W0803 17:02:44.466877   18808 eviction_manager.go:142] Failed to admit pod nvidia-device-plugin-daemonset-ldwsr_kube-system(eae3b613-96fb-11e8-80c4-fa163eb0739b) - node has conditions: [DiskPressure]
Aug 03 17:02:45 gpu16c-0001.novalocal kubelet[18808]: W0803 17:02:45.069755   18808 eviction_manager.go:142] Failed to admit pod calico-node-fjd28_kube-system(eb3fac9b-96fb-11e8-80c4-fa163eb0739b) - node has conditions: [DiskPressure]
Aug 03 17:02:45 gpu16c-0001.novalocal kubelet[18808]: W0803 17:02:45.670006   18808 eviction_manager.go:142] Failed to admit pod kube-proxy-kx8h6_kube-system(eb9b4b78-96fb-11e8-80c4-fa163eb0739b) - node has conditions: [DiskPressure]
Aug 03 17:02:45 gpu16c-0001.novalocal kubelet[18808]: W0803 17:02:45.776605   18808 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Aug 03 17:02:45 gpu16c-0001.novalocal kubelet[18808]: E0803 17:02:45.776713   18808 kubelet.go:2120] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Aug 03 17:02:46 gpu16c-0001.novalocal kubelet[18808]: W0803 17:02:46.269689   18808 eviction_manager.go:142] Failed to admit pod nvidia-device-plugin-daemonset-8h766_kube-system(ebf63f35-96fb-11e8-80c4-fa163eb0739b) - node has conditions: [DiskPressure]
Aug 03 17:02:46 gpu16c-0001.novalocal kubelet[18808]: W0803 17:02:46.868573   18808 eviction_manager.go:142] Failed to admit pod calico-node-c8rw9_kube-system(ec521274-96fb-11e8-80c4-fa163eb0739b) - node has conditions: [DiskPressure]
Aug 03 17:02:47 gpu16c-0001.novalocal kubelet[18808]: W0803 17:02:47.466671   18808 eviction_manager.go:142] Failed to admit pod nvidia-device-plugin-daemonset-ppkmf_kube-system(ecad92a7-96fb-11e8-80c4-fa163eb0739b) - node has conditions: [DiskPressure]
Aug 03 17:02:47 gpu16c-0001.novalocal kubelet[18808]: W0803 17:02:47.867567   18808 eviction_manager.go:142] Failed to admit pod kube-proxy-56djp_kube-system(ecea4b63-96fb-11e8-80c4-fa163eb0739b) - node has conditions: [DiskPressure]

```
but the disk no this node is fine 
```
[root@gpu16c-0001 overlay2]# df -h
Filesystem                                            Size  Used Avail Use% Mounted on
/dev/mapper/euleros-root                               35G   31G  3.1G  91% /
devtmpfs                                               63G     0   63G   0% /dev
tmpfs                                                  63G     0   63G   0% /dev/shm
tmpfs                                                  63G   17M   63G   1% /run
tmpfs                                                  63G     0   63G   0% /sys/fs/cgroup
/dev/nvme0n1                                          734G   25G  672G   4% /opt/avatar
/dev/vda1                                             477M  134M  314M  30% /boot
sfs-nas1.cn.com:/share-e4aaf55b  2.0T  236G  1.8T  12% /opt/avatar/sfs
tmpfs                                                  13G     0   13G   0% /run/user/0

[root@gpu16c-0001 overlay2]# df -Hi
Filesystem                                           Inodes IUsed IFree IUse% Mounted on
/dev/mapper/euleros-root                               2.4M  202k  2.2M    9% /
devtmpfs                                                17M   442   17M    1% /dev
tmpfs                                                   17M     1   17M    1% /dev/shm
tmpfs                                                   17M   504   17M    1% /run
tmpfs                                                   17M    14   17M    1% /sys/fs/cgroup
/dev/nvme0n1                                            49M  174k   49M    1% /opt/avatar
/dev/vda1                                              129k   346  128k    1% /boot
sfs-nas1.cn.com:/share-e4aaf55b   2.7G   21M  2.7G    1% /opt/avatar/sfs
tmpfs                                                   17M     1   17M    1% /run/user/0

```

here is docker info
```
[root@gpu16c-0001 overlay2]# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 10
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: nvidia runc
Default Runtime: nvidia
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871-dirty (expected: 4fc53a81fb7c994640722ac585fa9ca548971871)
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-327.62.59.83.h82.x86_64
Operating System: EulerOS 2.0 (SP2)
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 125.7GiB
Name: gpu16c-0001.novalocal
ID: SXW7:WSAM:PSUE:2AP3:4UU4:XI3N:UXBG:YFYX:Y4IG:6E4T:C4AZ:DZ6T
Docker Root Dir: /opt/avatar/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false


```

**What you expected to happen**:

node status not into DiskPressure when disk space is plenty.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.6", GitCommit:"9f8ebd171479bec0ada837d7ee641dec2f8c6dd1", GitTreeState:"clean", BuildDate:"2018-03-21T15:21:50Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.6", GitCommit:"9f8ebd171479bec0ada837d7ee641dec2f8c6dd1", GitTreeState:"clean", BuildDate:"2018-03-21T15:13:31Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
```
NAME="EulerOS"
VERSION="2.0 (SP2)"
ID="euleros"
ID_LIKE="rhel fedora centos"
VERSION_ID="2.0"
PRETTY_NAME="EulerOS 2.0 (SP2)"
ANSI_COLOR="0;31
```
- Kernel (e.g. `uname -a`):
Linux offlinetraining-0001.novalocal 3.10.0-327.59.59.46.h38.x86_64 #1 SMP Mon Dec 11 19:34:45 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

- Install tools:
kubeadm
- Others:

**Comment by yue9944882** [comment]
related #60558

it's likely caused by cadvisor

**Comment by tingweiwu** [comment]
@yue9944882 what do you means "caused by cadvisor" and how to solve it?  my Docker Root Dir  is /opt/avatar/docker.    and /opt/avatar still have 672G free and / hava 3G free.  I saw #60558. how can i find out which device my imageFs and nodeFs mounts on ?

**Comment by tingweiwu** [comment]
@yue9944882 I don't know if my guess is right or not. 
As described in kubernetes docs, the nodefs  is which /var/lib/kubelet mount and imagesfs is which Docker Root Dir mount?
```
The nodefs filesystem that kubelet uses for volumes, daemon logs, etc.
The imagefs filesystem that container runtimes uses for storing images and container writable layers.
```

**Comment by hugochinchilla** [comment]
I've found that kubelet is running with `/var/lib/kubelet` as root dir, I think it should be using `/var/vcap/data/kubelet` instead.

**Comment by nladuguie** [comment]
Hi, any ideas about the origin of this problem ?
We also have some nodes running into disk pressure, but not coming back from this state.
Our FS is not over used so we don't really understand what is going on.

**Comment by arun2arunraj** [comment]
hi All,

We are also facing this problem. How to overcome from this state. Any help would be much appreciated.

**Comment by sharkguto** [comment]
+1

**Comment by zrbilvoe** [comment]
+1

**Comment by joelsmith** [comment]
The original report looks like it's not a bug unless a threshold lower than the default 10% was specified.  If you look at
https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/
you'll see that `nodefs.available<10%` and `nodefs.inodesFree<5%` are among the defaults for the kubelet's `--eviction-hard` option, which means that if whichever volume contains /var/lib/kubelet gets to <10% free, the kubelet will exhibit a `DiskPressure` condition.  Same goes for inodes (as viewable via `df -i`) below 5% free.  The `df` output above shows only 9% free, so according to the configured thresholds, the node is under disk pressure. If you feel like those thresholds are too strict, you are welcome to configure your nodes with alternate thresholds (either in absolute terms or percentage terms). 

Absent any evidence that the kubelet is misbehaving, I'm going to close this one. If somebody has a node where its nodefs or imagefs usage (for bytes and inodes) are above the configured eviction thresholds, but it's still showing disk pressure, please send logs and info and re-open (or open a new one).

/close

**Comment by vaisov** [comment]
I did set `imagefs.available: 2Gi` in `/var/lib/kubelet/config.yaml`, but I still see errors in logs like:
`
Apr 23 15:16:30 y49295 kubelet[29149]: I0423 15:16:30.654916   29149 image_gc_manager.go:300] [imageGCManager]: Disk usage on image filesystem is at 89% which is over the high threshold (85%). Trying to free 9053972070 bytes down to the low threshold (80%).
Apr 23 15:16:30 y49295 kubelet[29149]: E0423 15:16:30.691238   29149 kubelet.go:1304] Image garbage collection failed multiple times in a row: failed to garbage collect required amount of images. Wanted to free 9053972070 bytes, but freed 0 bytes
`

I even tried to set evictionSoft limit, so my `/var/lib/kubelet/config.yaml` looks like:
`
evictionSoft:
  imagefs.available: 4Gi
evictionSoftGracePeriod:
  imagefs.available: 1m30s
evictionHard:
  imagefs.available: 2Gi
  memory.available: 100Mi
  nodefs.available: 10%
  nodefs.inodesFree: 5%
`

This didn't help either. Any ideas where it takes this 15% threshold from?

**Comment by mstrcnvs** [comment]
Hi there @joelsmith 
I am having a similar problem and this part caught my attention:

> The df output above shows only 9% free, so according to the configured thresholds, the node is under disk pressure.  

Reading the `df -Hi` output I can see "IUse%" as 9%, from my understanding this means that 9% is being used, not 9% is free (91% is free indeed).

We may be facing something similar on our project, random DiskPressure alerts without a clear cause. I am trying to monitor `df -i` while executing common processes we use on a daily basis, but without luck.

**Comment by joelsmith** [comment]
@mstrcnvs in the original report, it was the bytes used, not inodes used, which triggered the evictions. I was referring to this part of the original report:

```
[root@gpu16c-0001 overlay2]# df -h
Filesystem                                            Size  Used Avail Use% Mounted on
/dev/mapper/euleros-root                               35G   31G  3.1G  91% /
```
As you can see, the `/` volume only had 9% free.

**Comment by mstrcnvs** [comment]
Ah ok, understood. 

In our case neither one is at high usage.

```
$ df -h
Filesystem      Size  Used Avail Use% Mounted on
/dev/root       1.2G  476M  746M  39% /
devtmpfs        3.7G     0  3.7G   0% /dev
tmpfs           3.7G     0  3.7G   0% /dev/shm
tmpfs           3.7G  6.1M  3.7G   1% /run
tmpfs           3.7G     0  3.7G   0% /sys/fs/cgroup
tmpfs           3.7G     0  3.7G   0% /tmp
tmpfs           256K     0  256K   0% /mnt/disks
/dev/sda8        12M   28K   12M   1% /usr/share/oem
/dev/sda1        95G   18G   77G  19% /mnt/stateful_partition
tmpfs           1.0M  132K  892K  13% /var/lib/cloud
overlayfs       1.0M  176K  848K  18% /etc
$ df -Hi
Filesystem     Inodes IUsed IFree IUse% Mounted on
/dev/root         80k   15k   65k   19% /
devtmpfs         957k   210  957k    1% /dev
tmpfs            958k     1  958k    1% /dev/shm
tmpfs            958k  2.6k  955k    1% /run
tmpfs            958k    17  958k    1% /sys/fs/cgroup
tmpfs            958k    10  958k    1% /tmp
tmpfs            958k     1  958k    1% /mnt/disks
/dev/sda8        4.1k    11  4.1k    1% /usr/share/oem
/dev/sda1        6.3M  2.0M  4.4M   31% /mnt/stateful_partition
tmpfs            958k    45  958k    1% /var/lib/cloud
overlayfs        958k    75  958k    1% /etc
```

Yesterday we received "DiskPressure" health check alert in our only node and it lasted for approximately 20 mins. The node started to evict pods and such.

I am suspecting the inode usage got a peak but I can't find why. 

Thank you for the attention.

**Comment by Allen-yan** [comment]
I have the same  issue. My host disk  /inode usage  are both under 20%, 
but container warning Disk Pressure

**Comment by AliMD** [comment]
For other person ...
This two step solved my issue
1. Increase `addon_job_timeout`
2. Check node free space (at lease 15%)

In my case, one of the nodes had `DiskPressure` state

**Comment by michael-pratt** [comment]
I'm having this same problem with Kubernetes 1.17.4 and Redhat 7.

Kubelet is evicting docker images to free space:

    Dec 29 18:00:14 hostname kubelet: I1229 18:00:14.548513   13836 image_gc_manager.go:300] [imageGCManager]: Disk usage on image filesystem is at 85% which is over the high threshold (85%). Trying to free 2160300032 bytes down to the low threshold (80%).

Kubelet is responding to disk usage on `/`, which happens to be where Kubelet is installed:

    $ df -k /
    Filesystem     1K-blocks     Used Available Use% Mounted on
    /dev/nvme0n1p2  52403200 44032936   8370264  85% /

The problem is that Docker uses a different partition (`/scratch`)  for images, and it has plenty of space:

    $ docker info
    ...
    Docker Root Dir: /scratch/docker
    
    $ df -k /scratch
    Filesystem     1K-blocks      Used Available Use% Mounted on
    /dev/sda1      976283900 234476968 741806932  25% /scratch

Is there no way to tell Kubelet which partition to monitor? This gets us into a really bad state where our images are being evicted almost as fast as we can build them.

**Comment by chance2021** [comment]
Same issue. Run `docker system prune -a` to clean up some space taking by docker and then the node is running fine

**Comment by mostafa8026** [comment]
Some beginner question:
How can I change the eviction defaults? Is there any resources to read more? thanks

**Comment by therzv** [comment]
@mostafa8026 
[docs](https://kubernetes.io/docs/concepts/architecture/garbage-collection/)

you can always set 
```
HighThresholdPercent
LowThresholdPercent
```

**Comment by wotkddl21** [comment]
Did you check the volume of /var/lib/kubelet/pods ?
The actual space for the container is /var/lib/kubelet/pods/*

**Comment by mostafa8026** [comment]
For those who may find this issue, I found a workaround solution for myself. but someone else might see it helpful:

I'm using `containerd` and the main reason of low disk space is the images that are not used by any container. I've created a simple crontab to remove ununsed images every night:

```
0 3 * * * docker system prune -a -f && crictl --runtime-endpoint=unix:///run/containerd/containerd.sock rmi --prun
```

It also has the docker related command for image GC.
